File: src/lib/spotify.ts
import { cookies } from 'next/headers'

const SPOTIFY_API_BASE = 'https://api.spotify.com/v1'
const SPOTIFY_AUTH_BASE = 'https://accounts.spotify.com'

export type Track = {
  id: string
  name: string
  artists: Array<{ name: string }>
  album: {
    name: string
    images: Array<{ url: string; width: number; height: number }>
  }
  preview_url: string | null
  external_urls: { spotify: string }
  duration_ms: number
}

export type Artist = {
  id: string
  name: string
  external_urls: { spotify: string }
}

export type RecommendationParams = {
  seed_artists?: string[]
  seed_tracks?: string[]
  limit?: number
  min_energy?: number
  max_energy?: number
  target_energy?: number
  min_valence?: number
  max_valence?: number
  target_valence?: number
  min_danceability?: number
  max_danceability?: number
  target_danceability?: number
  min_tempo?: number
  max_tempo?: number
  target_tempo?: number
}

async function refreshAccessToken(refreshToken: string): Promise<string> {
  const clientId = process.env.SPOTIFY_CLIENT_ID
  const clientSecret = process.env.SPOTIFY_CLIENT_SECRET

  if (!clientId || !clientSecret) {
    throw new Error('Missing Spotify credentials')
  }

  const response = await fetch(`${SPOTIFY_AUTH_BASE}/api/token`, {
    method: 'POST',
    headers: {
      'Content-Type': 'application/x-www-form-urlencoded',
      Authorization: `Basic ${Buffer.from(`${clientId}:${clientSecret}`).toString('base64')}`,
    },
    body: new URLSearchParams({
      grant_type: 'refresh_token',
      refresh_token: refreshToken,
    }).toString(),
    cache: 'no-store',
  })

  if (!response.ok) {
    throw new Error(`Token refresh failed: ${response.statusText}`)
  }

  const data = (await response.json()) as { access_token: string; expires_in: number }
  return data.access_token
}

async function getAccessToken(): Promise<string> {
  const cookieStore = await cookies()
  const accessToken = cookieStore.get('spotify_access_token')?.value
  const refreshToken = cookieStore.get('spotify_refresh_token')?.value

  if (!accessToken || !refreshToken) {
    throw new Error('No Spotify tokens found')
  }

  // TODO: check expiry from cookie, refresh if needed
  // For now, assume token is fresh; production should check expiry
  return accessToken
}

async function spotifyFetch<T>(
  endpoint: string,
  options: RequestInit = {}
): Promise<T> {
  const token = await getAccessToken()

  const response = await fetch(`${SPOTIFY_API_BASE}${endpoint}`, {
    ...options,
    headers: {
      Authorization: `Bearer ${token}`,
      'Content-Type': 'application/json',
      ...options.headers,
    },
    cache: 'no-store',
  })

  if (!response.ok) {
    if (response.status === 401) {
      throw new Error('Spotify token expired')
    }
    throw new Error(`Spotify API error: ${response.statusText}`)
  }

  return response.json() as Promise<T>
}

export async function getRecommendations(
  params: RecommendationParams
): Promise<{ tracks: Track[] }> {
  const searchParams = new URLSearchParams()

  if (params.seed_artists) {
    searchParams.set('seed_artists', params.seed_artists.join(','))
  }
  if (params.seed_tracks) {
    searchParams.set('seed_tracks', params.seed_tracks.join(','))
  }
  searchParams.set('limit', String(params.limit || 20))

  if (params.target_energy !== undefined) {
    searchParams.set('target_energy', String(params.target_energy))
  }
  if (params.target_valence !== undefined) {
    searchParams.set('target_valence', String(params.target_valence))
  }
  if (params.target_danceability !== undefined) {
    searchParams.set('target_danceability', String(params.target_danceability))
  }
  if (params.target_tempo !== undefined) {
    searchParams.set('target_tempo', String(params.target_tempo))
  }

  if (params.min_energy !== undefined) {
    searchParams.set('min_energy', String(params.min_energy))
  }
  if (params.max_energy !== undefined) {
    searchParams.set('max_energy', String(params.max_energy))
  }
  if (params.min_valence !== undefined) {
    searchParams.set('min_valence', String(params.min_valence))
  }
  if (params.max_valence !== undefined) {
    searchParams.set('max_valence', String(params.max_valence))
  }
  if (params.min_danceability !== undefined) {
    searchParams.set('min_danceability', String(params.min_danceability))
  }
  if (params.max_danceability !== undefined) {
    searchParams.set('max_danceability', String(params.max_danceability))
  }
  if (params.min_tempo !== undefined) {
    searchParams.set('min_tempo', String(params.min_tempo))
  }
  if (params.max_tempo !== undefined) {
    searchParams.set('max_tempo', String(params.max_tempo))
  }

  return spotifyFetch<{ tracks: Track[] }>(
    `/recommendations?${searchParams.toString()}`
  )
}

export async function getTopTracks(limit = 5): Promise<{ items: Track[] }> {
  return spotifyFetch<{ items: Track[] }>(
    `/me/top/tracks?limit=${limit}&time_range=medium_term`
  )
}

export async function getTopArtists(limit = 5): Promise<{ items: Artist[] }> {
  return spotifyFetch<{ items: Artist[] }>(
    `/me/top/artists?limit=${limit}&time_range=medium_term`
  )
}

export async function getCurrentUser(): Promise<{ id: string; display_name: string }> {
  return spotifyFetch<{ id: string; display_name: string }>('/me')
}

export function buildAuthUrl(state: string, codeChallenge: string): string {
  const clientId = process.env.SPOTIFY_CLIENT_ID
  if (!clientId) throw new Error('Missing SPOTIFY_CLIENT_ID')

  const callbackUrl = process.env.SPOTIFY_CALLBACK_URL || 'https://teemu.space/callback'

  const params = new URLSearchParams({
    client_id: clientId,
    response_type: 'code',
    redirect_uri: callbackUrl,
    scope: 'user-top-read user-read-recently-played',
    state,
    code_challenge: codeChallenge,
    code_challenge_method: 'S256',
  })

  return `${SPOTIFY_AUTH_BASE}/authorize?${params.toString()}`
}

export async function exchangeAuthCode(
  code: string,
  codeVerifier: string
): Promise<{ access_token: string; refresh_token: string; expires_in: number }> {
  const clientId = process.env.SPOTIFY_CLIENT_ID
  if (!clientId) throw new Error('Missing SPOTIFY_CLIENT_ID')

  const callbackUrl = process.env.SPOTIFY_CALLBACK_URL || 'https://teemu.space/callback'

  const response = await fetch(`${SPOTIFY_AUTH_BASE}/api/token`, {
    method: 'POST',
    headers: { 'Content-Type': 'application/x-www-form-urlencoded' },
    body: new URLSearchParams({
      client_id: clientId,
      grant_type: 'authorization_code',
      code,
      redirect_uri: callbackUrl,
      code_verifier: codeVerifier,
    }).toString(),
    cache: 'no-store',
  })

  if (!response.ok) {
    const errorData = await response.text()
    console.error('Spotify token exchange error:', {
      status: response.status,
      statusText: response.statusText,
      body: errorData,
      callbackUrl,
    })
    throw new Error(`Auth exchange failed: ${response.statusText} - ${errorData}`)
  }

  return response.json() as Promise<{
    access_token: string
    refresh_token: string
    expires_in: number
  }>
}

File: backend/app/spotify.py
from __future__ import annotations

import base64
import asyncio
import os
import random
import time
import logging
from dataclasses import dataclass, field
from typing import Any

import httpx

from app.models import Album, AlbumImage, Artist, ExternalUrls, RecommendationsResponse, Track
from app.moods import search_queries_for_mood

SPOTIFY_API_BASE = "https://api.spotify.com/v1"
SPOTIFY_AUTH_URL = "https://accounts.spotify.com/api/token"
REDIS_KEY = "spotify:refresh_token"
_CACHE_TTL = 300  # seconds — cache per-mood results to reduce Spotify API calls

logger = logging.getLogger("spotify_mood.spotify")


class SpotifyServiceError(RuntimeError):
    pass


class TokenStore:
    """Persists the Spotify user refresh token in Upstash Redis via HTTP REST API.
    Upstash rotates nothing — we rotate by writing the new token after every refresh.
    Falls back to SPOTIFY_INITIAL_REFRESH_TOKEN env var on first run if Redis is empty.
    """

    def __init__(self) -> None:
        self._url = os.environ.get("UPSTASH_REDIS_REST_URL", "").rstrip("/")
        self._token = os.environ.get("UPSTASH_REDIS_REST_TOKEN", "")
        self._configured = bool(self._url and self._token)

    async def get(self, client: httpx.AsyncClient) -> str | None:
        if not self._configured:
            return os.environ.get("SPOTIFY_INITIAL_REFRESH_TOKEN")

        resp = await client.get(
            f"{self._url}/get/{REDIS_KEY}",
            headers={"Authorization": f"Bearer {self._token}"},
        )
        if resp.status_code != 200:
            logger.warning("redis_get_failed status=%s", resp.status_code)
            return os.environ.get("SPOTIFY_INITIAL_REFRESH_TOKEN")

        value = resp.json().get("result")
        if not value:
            # Redis key empty — seed from env var and persist it
            seed = os.environ.get("SPOTIFY_INITIAL_REFRESH_TOKEN")
            if seed:
                logger.info("redis_seeding_initial_token")
                await self.set(client, seed)
            return seed

        return value

    async def set(self, client: httpx.AsyncClient, refresh_token: str) -> None:
        if not self._configured:
            return

        resp = await client.post(
            self._url,
            json=["SET", REDIS_KEY, refresh_token],
            headers={"Authorization": f"Bearer {self._token}"},
        )
        if resp.status_code != 200:
            logger.warning("redis_set_failed status=%s body=%s", resp.status_code, resp.text[:120])
        else:
            logger.info("redis_refresh_token_updated")


@dataclass
class AccessTokenCache:
    access_token: str | None = None
    expires_at: float = 0

    def valid(self) -> bool:
        return bool(self.access_token) and time.time() < self.expires_at - 60


@dataclass
class SpotifyClient:
    _http_client: httpx.AsyncClient | None = field(default=None, repr=False)
    _app_cache: AccessTokenCache = field(default_factory=AccessTokenCache)
    _user_cache: AccessTokenCache = field(default_factory=AccessTokenCache)
    _token_store: TokenStore = field(default_factory=TokenStore)
    _mood_cache: dict[str, tuple[float, RecommendationsResponse]] = field(default_factory=dict)

    async def recommendations_for_mood(self, mood: str) -> RecommendationsResponse:
        cached = self._mood_cache.get(mood)
        if cached and time.time() < cached[0] + _CACHE_TTL:
            logger.info("mood=%s cache_hit", mood)
            return cached[1]
        result = await self._fetch_mood_recommendations(mood)
        self._mood_cache[mood] = (time.time(), result)
        return result

    async def _fetch_mood_recommendations(self, mood: str) -> RecommendationsResponse:
        queries = search_queries_for_mood(mood)
        query_one, query_two = random.sample(queries, 2)
        logger.info("mood=%s queries=%r", mood, [query_one, query_two])

        top_tracks, liked_songs, search_one, search_two = await self._gather_tracks(query_one, query_two)

        # Blend top tracks + liked songs, dedupe by id, filter unplayable
        personal_pool = {track.id: track for track in top_tracks + liked_songs}
        personal_candidates = [t for t in personal_pool.values() if t.preview_url]
        personal = random.sample(personal_candidates, min(8, len(personal_candidates)))
        seen = {track.id for track in personal}

        # General: filter preview_url, dedupe against personal
        general_candidates = search_one + search_two
        random.shuffle(general_candidates)
        general: list[Track] = []
        for track in general_candidates:
            if not track.preview_url:
                continue
            if track.id in seen:
                continue
            seen.add(track.id)
            general.append(track)
            if len(general) == 10:
                break

        logger.info("mood=%s general=%s personal=%s", mood, len(general), len(personal))
        return RecommendationsResponse(mood=mood, general=general, personal=personal)

    async def _gather_tracks(
        self, query_one: str, query_two: str
    ) -> tuple[list[Track], list[Track], list[Track], list[Track]]:
        return await self._with_client(
            lambda client: self._fetch_recommendation_sources(client, query_one, query_two)
        )

    async def _fetch_recommendation_sources(
        self,
        client: httpx.AsyncClient,
        query_one: str,
        query_two: str,
    ) -> tuple[list[Track], list[Track], list[Track], list[Track]]:
        search_one_response, search_two_response = await self._request_many(
            client,
            [
                ("/search", {"q": query_one, "type": "track", "limit": "20"}),
                ("/search", {"q": query_two, "type": "track", "limit": "10"}),
            ],
        )

        search_one = [
            sanitize_track(t)
            for t in search_one_response.get("tracks", {}).get("items", [])
        ]
        search_two = [
            sanitize_track(t)
            for t in search_two_response.get("tracks", {}).get("items", [])
        ]

        top_tracks: list[Track] = []
        liked_songs: list[Track] = []
        try:
            user_token = await self._user_access_token(client)
            top_resp, liked_resp = await asyncio.gather(
                client.get(
                    f"{SPOTIFY_API_BASE}/me/top/tracks",
                    params={"limit": "50", "time_range": "medium_term"},
                    headers={"Authorization": f"Bearer {user_token}"},
                ),
                client.get(
                    f"{SPOTIFY_API_BASE}/me/tracks",
                    params={"limit": "50"},
                    headers={"Authorization": f"Bearer {user_token}"},
                ),
            )
            if top_resp.status_code == 200:
                top_tracks = [sanitize_track(t) for t in top_resp.json().get("items", [])]
            else:
                logger.warning("top_tracks_failed status=%s", top_resp.status_code)
            if liked_resp.status_code == 200:
                liked_songs = [
                    sanitize_track(item["track"])
                    for item in liked_resp.json().get("items", [])
                    if item.get("track")
                ]
            else:
                logger.warning("liked_songs_failed status=%s", liked_resp.status_code)
        except SpotifyServiceError as exc:
            logger.warning("user_tracks_skipped reason=%s", exc)
        except Exception:
            logger.exception("user_tracks_fetch_failed")

        return top_tracks, liked_songs, search_one, search_two

    async def _user_access_token(self, client: httpx.AsyncClient) -> str:
        if self._user_cache.valid():
            return self._user_cache.access_token or ""

        client_id = os.environ.get("SPOTIFY_CLIENT_ID")
        client_secret = os.environ.get("SPOTIFY_CLIENT_SECRET")
        if not client_id or not client_secret:
            raise SpotifyServiceError("Missing SPOTIFY_CLIENT_ID / SPOTIFY_CLIENT_SECRET")

        refresh_token = await self._token_store.get(client)
        if not refresh_token:
            raise SpotifyServiceError("No refresh token in store")

        credentials = base64.b64encode(f"{client_id}:{client_secret}".encode()).decode()
        response = await client.post(
            SPOTIFY_AUTH_URL,
            data={"grant_type": "refresh_token", "refresh_token": refresh_token},
            headers={
                "Content-Type": "application/x-www-form-urlencoded",
                "Authorization": f"Basic {credentials}",
            },
        )

        if response.status_code >= 400:
            logger.warning("user_token_refresh_failed status=%s body=%s", response.status_code, response.text[:240])
            raise SpotifyServiceError("User token refresh failed")

        payload = response.json()
        access_token = payload.get("access_token")
        if not access_token:
            raise SpotifyServiceError("No access_token in refresh response")

        # Always persist the new refresh token — Spotify rotates it every time
        if new_refresh := payload.get("refresh_token"):
            await self._token_store.set(client, new_refresh)

        self._user_cache.access_token = access_token
        self._user_cache.expires_at = time.time() + int(payload.get("expires_in", 3600))
        return access_token

    async def _request_many(
        self,
        client: httpx.AsyncClient,
        requests: list[tuple[str, dict[str, str]]],
    ) -> list[dict[str, Any]]:
        access_token = await self._app_access_token(client)

        responses = await asyncio.gather(
            *[
                client.get(
                    f"{SPOTIFY_API_BASE}{path}",
                    params=params,
                    headers={"Authorization": f"Bearer {access_token}"},
                )
                for path, params in requests
            ]
        )

        data: list[dict[str, Any]] = []
        for response, (path, _) in zip(responses, requests):
            logger.info("spotify_api path=%s status=%s", path, response.status_code)
            if response.status_code == 401:
                self._app_cache.access_token = None
                raise SpotifyServiceError("Spotify authorization failed")
            if response.status_code >= 400:
                logger.warning("spotify_api_failed path=%s status=%s body=%s", path, response.status_code, response.text[:240])
                raise SpotifyServiceError("Spotify API request failed")
            data.append(response.json())
        return data

    async def _app_access_token(self, client: httpx.AsyncClient) -> str:
        if self._app_cache.valid():
            return self._app_cache.access_token or ""

        client_id = os.environ.get("SPOTIFY_CLIENT_ID")
        client_secret = os.environ.get("SPOTIFY_CLIENT_SECRET")
        if not client_id or not client_secret:
            raise SpotifyServiceError("Spotify credentials are not configured")

        credentials = base64.b64encode(f"{client_id}:{client_secret}".encode()).decode()
        response = await client.post(
            SPOTIFY_AUTH_URL,
            data={"grant_type": "client_credentials"},
            headers={
                "Content-Type": "application/x-www-form-urlencoded",
                "Authorization": f"Basic {credentials}",
            },
        )
        if response.status_code >= 400:
            logger.warning("app_token_failed status=%s body=%s", response.status_code, response.text[:240])
            raise SpotifyServiceError("Spotify app token request failed")

        payload = response.json()
        access_token = payload.get("access_token")
        if not access_token:
            raise SpotifyServiceError("No access_token in client_credentials response")

        self._app_cache.access_token = access_token
        self._app_cache.expires_at = time.time() + int(payload.get("expires_in", 3600))
        return access_token

    async def _with_client(self, operation):
        if self._http_client:
            return await operation(self._http_client)
        async with httpx.AsyncClient(timeout=10) as client:
            return await operation(client)


def sanitize_track(raw_track: dict[str, Any]) -> Track:
    album = raw_track.get("album") or {}
    external_urls = raw_track.get("external_urls") or {}

    return Track(
        id=str(raw_track.get("id") or ""),
        name=str(raw_track.get("name") or ""),
        artists=[
            Artist(name=str(artist.get("name") or ""))
            for artist in raw_track.get("artists", [])
        ],
        album=Album(
            name=str(album.get("name") or ""),
            images=[
                AlbumImage(
                    url=str(image.get("url") or ""),
                    width=image.get("width"),
                    height=image.get("height"),
                )
                for image in album.get("images", [])
            ],
        ),
        preview_url=raw_track.get("preview_url"),
        external_urls=ExternalUrls(spotify=str(external_urls.get("spotify") or "")),
        duration_ms=int(raw_track.get("duration_ms") or 0),
    )
